File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/web/security/CreatingUserAuthenticationSuccessHandler.java
package com.github.arhor.simple.expense.tracker.web.security;

import lombok.RequiredArgsConstructor;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.SavedRequestAwareAuthenticationSuccessHandler;
import org.springframework.stereotype.Component;

import com.github.arhor.simple.expense.tracker.service.UserService;

@Component
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class CreatingUserAuthenticationSuccessHandler extends SavedRequestAwareAuthenticationSuccessHandler {

    private final UserService userService;

    @Override
    public void onAuthenticationSuccess(
        final HttpServletRequest req,
        final HttpServletResponse res,
        final Authentication auth
    ) throws ServletException, IOException {

        userService.createNewUserIfNecessary(auth);
        super.onAuthenticationSuccess(req, res, auth);
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/web/security/CreatingUserAuthenticationSuccessHandler.kt
package com.github.arhor.simple.expense.tracker.web.security

import com.github.arhor.simple.expense.tracker.service.UserService
import org.springframework.security.core.Authentication
import org.springframework.security.web.DefaultRedirectStrategy
import org.springframework.security.web.authentication.SavedRequestAwareAuthenticationSuccessHandler
import org.springframework.stereotype.Component
import org.springframework.web.util.UriComponentsBuilder
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class CreatingUserAuthenticationSuccessHandler(
    private val service: UserService,
) : SavedRequestAwareAuthenticationSuccessHandler() {

    init {
        this.setUseReferer(true)
        this.redirectStrategy = object : DefaultRedirectStrategy() {
            override fun calculateRedirectUrl(contextPath: String?, url: String?): String {
                return super.calculateRedirectUrl(contextPath, url).let { path ->
                    if (path.endsWith("/sign-in")) {// TODO: use tha same constant as for WebSecurityConfig
                        path.let(UriComponentsBuilder::fromUriString)
                            .queryParam("success")
                            .build()
                            .toString()
                    } else {
                        path
                    }
                }
            }
        }
    }

    override fun onAuthenticationSuccess(req: HttpServletRequest, res: HttpServletResponse, auth: Authentication) {
        service.createNewUserIfNecessary(auth)
        super.onAuthenticationSuccess(req, res, auth)
    }
}
